avocado/utils/network/hosts.py

```diff
@@ -0,0 +1,115 @@
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#
+# See LICENSE for more details.
+#
+#
+# Copyright: 2019-2020 Red Hat Inc.
+# Authors : Beraldo Leal <bleal@redhat.com>
+
+"""
+This module provides an useful API for hosts in a network.
+"""
+
+from ..ssh import Session
+
+from .common import run_command
+from .interfaces import NetworkInterface
+from .exceptions import NWException
+
+
+class Host:
+    """This class represents a base Host and shouldn't be instantiated.
+
+    Use one of the child classes (LocalHost or RemoteHost).
+
+    During the initialization of a child, all interfaces will be detected and
+    available via `interfaces` attribute. This could be accessed on LocalHost
+    and RemoteHost instances.
+
+    So, for instance, you could have a local and a remote host::
+
+        remote = RemoteHost(host='foo', port=22,
+                            username='foo', password='bar')
+        local = LocalHost()
+
+    You can iterate over the network interfaces of any host::
+
+        for i in remote.interfaces:
+            print(i.name, i.is_link_up())
+    """
+    def __init__(self, host):
+        self.host = host
+
+    @property
+    def interfaces(self):
+        cmd = 'ls /sys/class/net'
+        try:
+            names = run_command(cmd, self).split()
+        except Exception as ex:
+            raise NWException("Failed to get interfaces: {}".format(ex))
+
+        return [NetworkInterface(if_name=name, host=self) for name in names]
+
+    def get_interface_by_ipaddr(self, ipaddr):
```

Comment:
@beraldoleal for this function getting error,
AttributeError: 'Host' object has no attribute 'remote_session'

Reply:
@smruti77, how are you calling it?

```python3
In [1]: from avocado.utils.network.hosts import LocalHost                                                                                     
In [2]: local = LocalHost()                                                                                                                   
In [3]: interface = local.get_interface_by_ipaddr('127.0.0.1')                                                                                
In [4]: interface.name                                                                                                                        
Out[4]: 'lo'
In [5]: interface.get_mtu()                                                                                                                   
Out[5]: 1500
```

Reply:
@beraldoleal  i am trying for the remote machine interface name,

from avocado.utils.network import hosts
remotehost = hosts.RemoteHost(self.peer_ip, self.peer_user, password=self.peer_password)
host = hosts.Host(remotehost)
interface = host.get_interface_by_ipaddr(self.peer_ip)
peer_interface = interface.name

Reply:
@smruti77, please, read the docstrings on `Host`, `LocalHost` and `RemoteHost`. You should not have instances of `Host`, only `LocalHost` and `RemoteHost`.

Reply:
Could we turn `Host` into a private class?

Reply:
I don't like much the underscore convention for classes. For me `class RemoteHost(_Host)`, it is super strange. But I agree that we have to hide this from the user.

For now, I'm going to check the class type on `__init__` and raise an exception. Not perfect, but it seems to work.

Reply:
@smruti77, just to make clear how to use this:

```python3
from avocado.utils.network.hosts import LocalHost, RemoteHost     
                                                                    
local = LocalHost() 
remote = RemoteHost('127.0.0.1', username='foo', port=22, password='bar') 

for interface in local.interfaces:
   print(interface.name)

for interface in remote.interfaces:
   print(interface.name)
```

Reply:
Or just define `__all__` in the module?

Reply:
The variable `__all__` is just a list of public objects of that module, as interpreted by `import *`, useful for tab completion and IDEs, right?

Reply:
Yes. Same effect as the leading underscore.

Reply:
> @smruti77, just to make clear how to use this:
> 
> ```python
> from avocado.utils.network.hosts import LocalHost, RemoteHost     
>                                                                     
> local = LocalHost() 
> remote = RemoteHost('10.0.0.1', username='foo', port=22, password='bar') 
> 
> for interface in local.interfaces:
>    print(interface.name)
> 
> for interface in remote.interfaces:
>    print(interface.name)
> ```

got it,yesterday i tried again it worked.